File: typescript/src/application/service/assetService.ts
import Decimal from "decimal.js";
import { Account, Stock } from "../../domain/stock.js";
import { StockSymbol } from "../../domain/stockSymbol.js";

export const AssetService = {
  evaluateStock(stock: Stock, prices: Map<StockSymbol, Decimal>): Decimal {
    const price = prices.get(stock.symbol);
    if (price === undefined) {
      throw new Error(`missing price for ${stock.symbol}`);
    }
    return stock.qty.times(price);
  },

  totalValuation(account: Account, prices: Map<StockSymbol, Decimal>): Decimal {
    return account.stocks
      .map((e) => AssetService.evaluateStock(e, prices))
      .reduce((acc, v) => acc.plus(v), new Decimal(0))
      .plus(account.cash);
  },
};

File: golang/internal/application/service/asset_service.go
package service

import (
	"fmt"

	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

func EvaluateStock(stock domain.Stock, prices map[domain.StockSymbol]decimal.Decimal) (decimal.Decimal, error) {
	price, ok := prices[stock.Symbol]
	if !ok {
		return decimal.Zero, fmt.Errorf("price not found for symbol: %s", stock.Symbol)
	}
	return stock.Qty.Mul(price), nil
}

func TotalValuation(account domain.Account, prices map[domain.StockSymbol]decimal.Decimal) (decimal.Decimal, error) {
	total := account.Cash
	for _, s := range account.Stocks {
		val, err := EvaluateStock(s, prices)
		if err != nil {
			return decimal.Zero, err
		}
		total = total.Add(val)
	}
	return total, nil
}
